### PR TITLE
realtek: add support for TP-Link SG2452P v4 aka T1600G-52PS v4

### PR DESCRIPTION
**Pull request by andyboeh**
This is an RTL8393-based switch with 802.3af on all 48 ports.

Specifications:
---------------
 * SoC:       Realtek RTL8393M
 * Flash:     32 MiB SPI flash
 * RAM:       256 MiB
 * Ethernet:  48x 10/100/1000 Mbps with PoE+
 * Buttons:   1x "Reset" button, 1x "Speed" button
 * UART:      1x serial header, unpopulated
 * PoE:       12x TI TPS23861 I2C PoE controller
 * SFP:       4 SFP ports

Works:
------
  - (48) RJ-45 ethernet ports
  - Switch functions
  - Buttons
  - All LEDs on front panel except port LEDs
  - Fan monitoring and basic control

Not yet enabled:
----------------
  - Power-over-Ethernet (driver available, but doesn't enable "auto" mode)
  - Port LEDs
  - SFP cages

Install via web interface:
-------------------------

Not supported at this time.

Install via serial console/tftp:
--------------------------------

The U-Boot firmware drops to a TP-Link specific "BOOTUTIL" shell at
38400 baud. There is no known way to exit out of this shell, and no
way to do anything useful.

Ideally, one would trick the bootloader into flashing the sysupgrade
image first. However, if the image exceeds 6MiB in size, it will not
work. To install OpenWRT:

Prepare a tftp server with:
 1. server address: 192.168.0.146
 2. the image as: "uImage.img"

Power on device, and stop boot by pressing any key.
Once the shell is active:
 1. Ground out the CLK (pin 16) of the ROM (U6)
 2. Select option "3. Start"
 3. Bootloader notes that "The kernel has been damaged!"
 4. Release CLK as soon as bootloader thinks image is corrupted.
 5. Bootloader enters automatic recovery -- details printed on console
 6. Watch as the bootloader flashes and boots OpenWRT.

Blind install via tftp:
-----------------------

This method works when it's not feasible to install a serial header.

Prepare a tftp server with:
 1. server address: 192.168.0.146
 2. the image as: "uImage.img"
 3. Watch network traffic (tcpdump or wireshark works)
 4. Power on the device.
 5. Wait 1-2 seconds then ground out the CLK (pin 16) of the ROM (U6)
 6. When 192.168.0.30 makes tftp requests, release pin 16
 7. Wait 2-3 minutes for device to auto-flash and boot OpenWRT

Signed-off-by: Andreas Böhler <dev@aboehler.at>

### CODE REVIEW

**Comment by andyboeh** [comment]
Some additional notes:

  * Currently, I cannot come up with SFP support as I do not have any SFP modules.
  * I did not enable the hwmon driver for the tps23861 on purpose: The driver does not initialize the hardware and is thus unusable. Upstream support seems very unlikely, so I'm currently working on a user space implementation.

File: target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts [hunk 1]
```diff
@@ -0,0 +1,330 @@
...
+		led-running = &led_sys;
+		led-upgrade = &led_sys;
+	};
+
+	chosen {
+		bootargs = "console=ttyS0,38400";
+	};
+
+	keys {
+		compatible = "gpio-keys-polled";
```
Comment (musashino205) [inline target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts:29]: The gpio driver of Realtek SoCs supports interrupt, so this can be replaced by ```gpio-keys```. Then, you can drop ```poll-interval``` property.

Comment (andyboeh) [reply target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts:29]: OK, I didn't know that. My reference for this was the GS1900-48 where `gpio-keys-polled` is used, but now I saw that the Panasonic uses `gpio-keys`.

Comment (musashino205) [reply target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts:29]: GS1900-48 uses the pin on RTL8231 for keys and it doesn't support interrupt. Panasonic Switch-M48eG PN28480K uses the pin on PCA9539 controller, it supports interrupt via SoC GPIO controller.

Comment (andyboeh) [reply target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts:29]: Oh, well, I should have had a closer look :) Thanks for the clarification!

File: target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts [hunk 2]
```diff
@@ -0,0 +1,330 @@
...
+		alarm-gpios = <&gpio0 7 GPIO_ACTIVE_LOW>;
+		gpios = <&gpio0 4 GPIO_ACTIVE_LOW>;
+		/* the actual speeds (rpm) are unknown, just use dummy values */
+		gpio-fan,speed-map = <1 0>, <2 1>;
+		#cooling-cells = <2>;
+	};
+
+	gpio_fan_psu_2 {
+		compatible = "gpio-fan";
+		alarm-gpios = <&gpio0 14 GPIO_ACTIVE_LOW>;
```
Comment (svanheule) [inline target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts:64]: Same as for `gpio_fan_sys`: does this fan run at a fixed speed, or is this a work-around for `alarm-gpios` on `gpio_fan_psu_1` only supporting one GPIO line?

File: target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts [hunk 3]
```diff
@@ -0,0 +1,330 @@
...
+	};
+
+	leds {
+		pinctrl-names = "default";
+		compatible = "gpio-leds";
+
+		led_sys: sys {
```
Comment (svanheule) [inline target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts:71]: Recent gpio-leds bindings require node names of the form `led-[0-9]+`, as well as the "color" and "function" properties. OpenWrt still needs the label, but can you use default function names from the  LED header? If those don't really match, perhaps you can add a comment when using a custom (string) value for "function", to clarify what it indicates.

File: target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts [hunk 4]
```diff
@@ -0,0 +1,330 @@
...
+		};
+	};
+
+	i2c-gpio-0 {
+		compatible = "i2c-gpio";
+		sda-gpios = <&gpio0 2 (GPIO_ACTIVE_HIGH | GPIO_OPEN_DRAIN)>;
+		scl-gpios = <&gpio0 1 (GPIO_ACTIVE_HIGH | GPIO_OPEN_DRAIN)>;
+		i2c-gpio,delay-us = <2>;
+		#address-cells = <1>;
+		#size-cells = <0>;
```
Comment (svanheule) [inline target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts:108]: No children on this I2C bus? If this is for the PSE controllers, you can (should) already add these, even if they cannot be used. That way you can document them (for someone else) to enable these in the future. If adding them enabled causes errors or warnings, add `status = "disabled"` and note this in the commit message.

Do how the I2C lines of the SFP cages are wired up?

File: target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts [hunk 5]
```diff
@@ -0,0 +1,330 @@
...
+
+	gpio-restart {
+		compatible = "gpio-restart";
+		gpios = <&gpio0 17 GPIO_ACTIVE_LOW>;
+	};
+
+	virtual_flash {
+		compatible = "mtd-concat";
+
+		devices = <&syspart &usrpart>;
```
Comment (svanheule) [inline target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts:119]: `devices = <&syspart>, <&usrpart>`

You are providing a list of MTD partitions, so the list should have MTD partition entries. The `<&ref_a &ref_b>` notation indicates that `&ref_b` is a property of `&ref_a`, which is not the case here. Once compiled, both notations end up being identical, but I don't think we should be abusing that implementation detail.

File: target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts [hunk 6]
```diff
@@ -0,0 +1,330 @@
...
+		line-name = "poe-enable";
+	};
+};
+
+&spi0 {
+	status = "okay";
+	flash@0 {
+		compatible = "jedec,spi-nor";
+		reg = <0>;
+		spi-max-frequency = <10000000>;
```
Comment (svanheule) [inline target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts:149]: Did you get this value from a datasheet?

File: target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts [hunk 7]
```diff
@@ -0,0 +1,330 @@
...
+				read-only;
+			};
+		};
+	};
+};
+
+&ethernet0 {
+	mdio: mdio-bus {
+		compatible = "realtek,rtl838x-mdio";
+		regmap = <&ethernet0>;
```
Comment (svanheule) [inline target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts:193]: `regmap` property is not used, and won't have any use in the future, so it can be removed. (Note to self: should clean up existing DTS files)

File: target/linux/realtek/image/rtl839x.mk [hunk 8]
```diff
@@ -11,6 +11,17 @@ define Device/panasonic_m48eg-pn28480k
 endef
 TARGET_DEVICES += panasonic_m48eg-pn28480k
 
+define Device/tplink_sg2452p-v4
+  SOC := rtl8393
+  KERNEL_SIZE := 6m
+  IMAGE_SIZE := 26m
+  DEVICE_VENDOR := TP-Link
+  DEVICE_MODEL := SG2452P
+  DEVICE_VARIANT := v4
```
Comment (svanheule) [inline target/linux/realtek/image/rtl839x.mk:20]: You could add `DEVICE_ALT0_*` variables to indicate the T1600G-52PS v4 is really the same device. Do you where  TP-Link uses these names? Or are both used as model names on the device label?

File: target/linux/realtek/rtl839x/config-5.10 [hunk 9]
```diff
@@ -137,6 +137,7 @@ CONFIG_MTD_SPLIT_EVA_FW=y
 CONFIG_MTD_SPLIT_FIRMWARE=y
 CONFIG_MTD_SPLIT_TPLINK_FW=y
 CONFIG_MTD_SPLIT_UIMAGE_FW=y
+CONFIG_MTD_VIRT_CONCAT=y
```
Comment (svanheule) [inline target/linux/realtek/rtl839x/config-5.10:140]: Please put this change in a separate commit. You're adding the first device that will use it, so that could the be rationale you can put in the commit message.

File: target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts [hunk 10]
```diff
@@ -0,0 +1,330 @@
...
+	gpio_fan_psu_1 {
+		pinctrl-names = "default";
+		pinctrl-0 = <&disable_jtag>;
+		compatible = "gpio-fan";
+
+		alarm-gpios = <&gpio0 7 GPIO_ACTIVE_LOW>;
+		gpios = <&gpio0 4 GPIO_ACTIVE_LOW>;
+		/* the actual speeds (rpm) are unknown, just use dummy values */
+		gpio-fan,speed-map = <1 0>, <2 1>;
+		#cooling-cells = <2>;
```
Comment (svanheule) [inline target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts:59]: How does stock FW perform thermal management? No temperature sensors on the board?

**Review by svanheule — COMMENTED** [review]
For the commit message, could you also add what the specified power ratings are for this switch? That should make it possible too for others to write changes that require this info.

File: target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts [hunk 11]
```diff
@@ -0,0 +1,330 @@
...
+		poll-interval = <20>;
+
+		reset {
+			label = "reset";
+			gpios = <&gpio0 20 GPIO_ACTIVE_LOW>;
+			linux,code = <KEY_RESTART>;
+		};
+
+		speed {
+			label = "speed";
```
[the remark this answers is not in the document]
Comment (andyboeh) [reply target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts:39]: It switches LED mode from displaying the speed to the PoE status; it's labelled "Speed" on the front panel.

File: target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts [hunk 12]
```diff
@@ -0,0 +1,330 @@
...
+			gpios = <&gpio0 19 GPIO_ACTIVE_LOW>;
+			linux,code = <BTN_0>;
+		};
+	};
+
+	gpio_fan_sys {
+		compatible = "gpio-fan";
+		alarm-gpios = <&gpio0 12 GPIO_ACTIVE_LOW>;
```
[the remark this answers is not in the document]
Comment (andyboeh) [reply target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts:47]: Yes

File: target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts (hunk 2 above)
Comment (andyboeh) [reply target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts:64]: Exactly - both PSU fans can be slowed down by the same GPIO line but have individual alarm lines.

File: target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts (hunk 4 above)
Comment (andyboeh) [reply target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts:108]: That's the PSE controllers. I'll add disabled nodes as the tps23861 driver is of no practical use.

Unfortunately, I was not able to figure out the SFP ports.

File: target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts (hunk 5 above)
Comment (andyboeh) [reply target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts:119]: That's a plain copy of the TL-SG2008P. In the meantime, however, I'm quite confident that we don't need it at all: the partitions are contiguous anyway, so I could just redefine the partition map and remove the `mtd-concat` driver altogether. Same applies to the TL-SG2008P.

File: target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts (hunk 6 above)
Comment (andyboeh) [reply target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts:149]: Nope, copy of TL-SG2008P.

File: target/linux/realtek/image/rtl839x.mk (hunk 8 above)
Comment (andyboeh) [reply target/linux/realtek/image/rtl839x.mk:20]: On the front panel it reads T1600G-52PS, on the bottom it reads both names. I opted for Tl-SG2452 to be consistent with the naming of the other TP-Link devices supported on the Realtek target.

File: target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts (hunk 10 above)
Comment (andyboeh) [reply target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts:59]: AFAIK only based on the PoE budget used. It's definitely done in user space as I found the respective function calls in the `core` binary. My Ghidra skills are limited, so I didn't trace it further.
There is also an even slower fan mode, but the Ghidra dump shows some weird GPIO settings that might be related to the LED shift register outputs.

File: target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts (hunk 5 above)
Comment (svanheule) [reply target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts:119]: @mrnuke Can you remind us why you used mtd-concat? IIRC you added the kernel and image size limits later, so maybe mtd-concat became unnecessary after that.

Comment (mrnuke) [reply target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts:119]: mtd-concat was the easiest way to make it compatible with the existing sysupgrade scripts. I thought about separate "kernel" and "rootfs" partitions, but I was too lazy to make sysupgrade work with it.

File: target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts [hunk 13]
```diff
@@ -0,0 +1,330 @@
...
+				reg = <0x100000 0x600000>;
+			};
+			usrpart: partition@700000 {
+				label = "usrimg1";
+				reg = <0x700000 0xa00000>;
+			};
+			partition@1100000 {
+				label = "usrimg2";
+				reg = <0x1100000 0xa00000>;
+			};
```
Comment (mrnuke) [inline target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts:176]: Can you use tplink_sg2xxx_fix_mtdparts() to extend usrimg1 and use the space of both partitions, like we did for 2008P?

https://github.com/openwrt/openwrt/blob/104de8abe455405d43133edc551d50645934d5be/target/linux/realtek/base-files/lib/upgrade/platform.sh#L12-L20

File: target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts (hunk 2 above)
Comment (svanheule) [reply target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts:64]: If all fans are controlled by the same set of GPIOs, then I think the devicetree description should be able to reflect this.

I think we could propose upstream to use [`devm_gpiod_get_array_optional()`](https://elixir.bootlin.com/linux/latest/source/include/linux/gpio/consumer.h#L102) to support multiple alarm lines for `alarm-gpios`. That call is also available in 5.10, so the change could probably be backported.

File: target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts (hunk 10 above)
Comment (svanheule) [reply target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts:59]: Can you list the GPIO manipulations (or raw register changes)? It would be interesting to see what's going on.

I assume the "slower fan mode" corresponds to control value "0"? Does control value "3" result in any noticeable change?

File: target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts (hunk 13 above)
Comment (andyboeh) [reply target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts:176]: I decided against it, because I don't want to overwrite the backup image. A kernel restore should then be sufficient to revert back to stock. Furthermore, I didn't want to change the default baud rate.

Comment (svanheule) [reply target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts:176]: Then `IMAGE_SIZE` in the build recipe should also be limited to 6M + 10MB, as I see it's currently 26MB.

Comment (svanheule) [reply target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts:176]: Can users just restore the kernel from a vendor upgrade file, or do they need to take special steps to extract this part? You could add this in the commit message, if you know the steps.

Comment (mrnuke) [reply target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts:176]: This procedure should be verified. I remember trying this on a SG2008P, and the boot process got stuck at "RSA Verify failed!".

Comment (andyboeh) [reply target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts:176]: I didn't try it, no - that's on my ToDo List. I don't think it will work with a vendor upgrade file, but it should work with the backup image I extracted from the IC. Maybe a vendor image could be prepared in a way it is compatible.

File: target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts (hunk 2 above)
Comment (andyboeh) [reply target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts:64]: Sounds like an option, but after my recent submission regarding the tps23861 driver I will not get in touch with anything kernel related in the near future.

File: target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts (hunk 10 above)
Comment (andyboeh) [reply target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts:59]: I'll send you a PM on the forum, you probably have more experience with that.

File: target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts (hunk 2 above)
Comment (svanheule) [reply target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts:64]: Upstream can be hard :-/

I've started with converting the binding to YAML and may have a look at multi-alarm-gpios after next week. We can keep the DT like you've suggested for now, although it should be documented that both fans are controlled by the GPIO values.

File: target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts (hunk 13 above)
Comment (andyboeh) [reply target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts:242]: Since it doesn't make any difference whether usrimg2 is used or not, I extended it to usrimg2 as well. I don't think it's necessary to tell U-Boot about it so I'm not modifying the U-Boot variables. I'll try to come up with a way back to stock - it should be possible to flash kernel + usrimg1 via the TFTP recovery method, at least the one extracted from my flash backup.

File: target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts (hunk 2 above)
Comment (robimarko) [reply target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts:66]: @andyboeh You cannot really think like that, upstreaming is hard and won't get easier, but is the only way to prevent maintaining local patches forever.

OpenWrt is full of never upstreamed hacks that have piled up over the years as people treated it way around dealing with upstream, you can probably see what an issue every kernel bump is.

Just look at the mess of a target Realtek is, its still stuck on 5.10

Comment (andyboeh) [reply target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts:66]: I know and I'm not against upstreaming, don't get me wrong here. It's just that I do not have the time and resources to do that. That's also one of the reasons why I deleted the tps23861 patches instead of including a local copy here.

It's hard enough for me to contribute to OpenWrt at all - I wanted a port to this switch model since I bought it in 2020 and just now I have a bit of time to do so. However, this period is limited and will be over or at least very limited in two weeks due to work and family obligations.

I'm very glad that people like you contribute much more to this project than I can do.

File: target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts (hunk 4 above)
Comment (svanheule) [reply target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts:110]: If you know how the TPS23861 outputs map to switch ports, that would also be interesting to note in a comment (per node).

File: target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts (hunk 2 above)
Comment (svanheule) [reply target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts:66]: We'll see how far we can get in that short time span then. Just try to document every little detail you know, that could be very helpful if others want to write changes when drivers are updated to provide more features.

Comment (robimarko) [reply target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts:66]: I understand time is usually the limiting factor anywhere.
Though, it looks like PoE PSE basic framework is slowly getting assembled so in future we could add 802.3at/af/bt support as well instead of having multiple userspace implementations.

File: target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts (hunk 13 above)
Comment (mrnuke) [reply target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts:242]: Once you decrypt the firmware, you can use the following script to separate out the image bits:

https://gist.github.com/mrnuke/2dbcc6afd40c8d5a819457dfff8cdbcf

Comment (andyboeh) [reply target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts:242]: I wrote a similar script literally just now! Mine includes the decryption, though. I'll publish mine once I've tested back-to-stock.

Comment (mrnuke) [reply target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts:242]: Battle of the scripts!

Comment (andyboeh) [reply target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts:242]: Back to stock works fine by simply running `sysupgrade -n -F` with the extracted parts from my script: https://gist.github.com/andyboeh/e5cf74d2912dee527823d654414e4323

File: target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts (hunk 5 above)
Comment (andyboeh) [reply target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts]: It works fine on my switch without `mtd-concat`.

File: target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts (hunk 4 above)
Comment (andyboeh) [reply target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts:110]: Not yet. I'll investigate this later.

**Comment by andyboeh** [comment]
Regarding SFP: I'm quite certain that i2c is not connected to the SoC from the SFP cages. The stock firmware nowhere displays information about the module and does not detect insertion/removal of the module - it only reports link changes.

File: target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts (hunk 4 above)
Comment (andyboeh) [reply target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts:110]: I re-enabled the hwmon driver, as @mrnuke pointed out that one simple i2cset operation will set all TPS23861 ICs to the required mode. Comments about the mapping are included now.

File: target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts (hunk 3 above)
Comment (andyboeh) [reply target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts]: Done

File: target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts [hunk 14]
```diff
@@ -0,0 +1,412 @@
...
+			gpios = <&gpio0 18 GPIO_ACTIVE_HIGH>;
+			color = <LED_COLOR_ID_GREEN>;
+			function = LED_FUNCTION_INDICATOR;
+		};
+
+		led-3 {
+			label = "green:speed";
+			gpios = <&gpio0 8 GPIO_ACTIVE_HIGH>;
+			color = <LED_COLOR_ID_GREEN>;
+			function = LED_FUNCTION_INDICATOR;
```
Comment (svanheule) [inline target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts:100]: If this LED and "green:poe" are used to indicate what the port LEDs are currently showing, then "function indicator" does seem the more appropriate selection here.

File: target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts [hunk 15]
```diff
@@ -0,0 +1,412 @@
...
+
+		led-4 {
+			label = "green:poe";
+			gpios = <&gpio0 9 GPIO_ACTIVE_HIGH>;
+			color = <LED_COLOR_ID_GREEN>;
+			function = LED_FUNCTION_INDICATOR;
+		};
+
+		led-5 {
+			label = "green:fan";
```
Comment (svanheule) [inline target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts:111]: Does this LED indicate "fans running at low speed"?

File: target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts [hunk 16]
```diff
@@ -0,0 +1,412 @@
...
+				reg = <0x0 0xe0000>;
+				read-only;
+			};
+			partition@e0000 {
+				label = "u-boot-env";
+				reg = <0xe0000 0x20000>;
+			};
+
+			/* We use the "sys", "usrimg1" and "usrimg2" partitions
+			   as firmware since the kernel needs to be in "sys", but the
```
Comment (svanheule) [inline target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts:238]: Please use the conventional kernel comment style, with * prefixing every comment line. It wasn't immediately clear to me that the three original partitions were actually in a comment.

**Review by svanheule — COMMENTED** [review]
Some minor comments on the update. If needed, I can also implement these changes when merging.

File: target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts [hunk 17]
```diff
@@ -0,0 +1,412 @@
...
+			gpios = <&gpio0 13 GPIO_ACTIVE_HIGH>;
+			color = <LED_COLOR_ID_GREEN>;
+			function = LED_FUNCTION_STATUS;
+		};
+
+		led-1 {
+			label = "amber:fan";
+			gpios = <&gpio0 16 GPIO_ACTIVE_HIGH>;
+			color = <LED_COLOR_ID_AMBER>;
+			function = LED_FUNCTION_INDICATOR;
```
[the remark this answers is not in the document]
Comment (andyboeh) [reply target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts:85]: I think this actually indicates that a fan is broken. I though about "LED_FUNCTION_ALARM", but then this LED is mutually exclusive with the green fan LED, and the green one is not an ALARM function.

File: target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts [hunk 18]
```diff
@@ -0,0 +1,412 @@
...
+			function-enumerator = <1>;
+		};
+
+		led-2 {
+			label = "green:poe-max";
+			gpios = <&gpio0 18 GPIO_ACTIVE_HIGH>;
+			color = <LED_COLOR_ID_GREEN>;
+			function = LED_FUNCTION_INDICATOR;
```
[the remark this answers is not in the document]
Comment (andyboeh) [reply target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts:93]: Yeah, I was unsure whether it's status or indicator ;)

[the remark this answers is not in the document]
Comment (andyboeh) [reply target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts:93]: It might, however, also be ALARM as it's used to indicate when PoE is over or near the budget (if I remember correctly).

File: target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts (hunk 14 above)
Comment (andyboeh) [reply target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts:100]: Exactly this is the case.

File: target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts (hunk 15 above)
Comment (andyboeh) [reply target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts:111]: I think it means "Fans OK" - see above, mutually exclusive with amber:fan

File: target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts [hunk 19]
```diff
@@ -0,0 +1,412 @@
...
+		led-4 {
+			label = "green:poe";
+			gpios = <&gpio0 9 GPIO_ACTIVE_HIGH>;
+			color = <LED_COLOR_ID_GREEN>;
+			function = LED_FUNCTION_INDICATOR;
+		};
+
+		led-5 {
+			label = "green:fan";
+			gpios = <&gpio0 15 GPIO_ACTIVE_HIGH>;
```
[the remark this answers is not in the document]
Comment (andyboeh) [reply target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts:112]: I'm unsure if I've sorted the LEDs at all - I'll sort them by GPIO number.

File: target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts (hunk 17 above)
[the remark this answers is not in the document]
Comment (svanheule) [reply target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts:85]: The [installation manual](https://static.tp-link.com/upload/manual/2022/202203/20220331/7106509248_T1600G_IG.pdf) does indeed state the following:

> FAN
> Green: All the fans work properly.
> Yellow: Not all the fans work properly.

Which does seem to indicate that there is one bi-directional LED, so "amber:fan" and "green:fan" cannot be enabled simultaneously (something to note in the commit message).

ALARM is be more applicable, but maybe FAULT matches best. Or just use `function = "alarm-poe"` and `function = "fault-fan"`, as otherwise the automatic names would end up being the very generic "green:alarm" for PoE overload, and "green:fault-0"/"amber:fault-1" for FAN fault.

**Comment by andyboeh** [comment]
I fixed the LED assignments and also corrected a mistake with the "compatible" string: I had `rtl838x-soc` instead of `rtl8393-soc`.

**Review by svanheule — COMMENTED** [review]
Some last minor nits on the comments, otherwise I think this patch is good to go.

File: target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts [hunk 20]
```diff
@@ -0,0 +1,410 @@
...
+				label = "u-boot-env";
+				reg = <0xe0000 0x20000>;
+			};
+
+			/* We use the "sys", "usrimg1" and "usrimg2" partitions
+			 *  as firmware since the kernel needs to be in "sys", but the
+			 *  partition is too small to hold the "rootfs" as well.
+			 *  The original partition map contains:
+
+			 *   partition@100000 {
```
[the remark this answers is not in the document]
Comment (mrnuke) [reply target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts:241]: If the intent is to document the vendor's partition table, why not write it out and use mtd-concat to create the `firmware` partition? It takes less lines of code to do so.

[the remark this answers is not in the document]
Comment (svanheule) [reply target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts:241]: If mtd-concat is not needed here, maybe we should remove it on the SG2008/SG2210 too. On other targets (e.g. ath79 TP-Link devices), concatenation of contiguous kernel/rootfs partitions is also done without mtd-concat, so I'm fine with this approach.

One benefit that the mtd-concat approach does have, is that it's explicit. But mtd-concat introduces an extra layer which is not needed here. The partitions are contiguous, so they can be manually concatenated. I'm fine with the original layout being documented in DT format.

[the remark this answers is not in the document]
Comment (mrnuke) [reply target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts:241]: Okay. I'll prepare a patch to remove mtd-concat on SG2008/SG2210.

Only disadvantage would be it's a little harder to mtd write the vendor partitions back -- which I don't care much about.

**Comment by svanheule** [comment]
Merged with commit 5f8c86e654 ("realtek: add support for TP-Link SG2452P v4 aka T1600G-52PS v4"). Thanks!

File: target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts (hunk 20 above)
[the remark this answers is not in the document]
Comment (svanheule) [reply target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts:241]: For safeloader devices, the image generation tool has an option to create sysupgrade images from vendor files. This takes the vendor kernel, pads it to the original kernel partition size, and then appends the vendor rootfs. In this case, I suppose a simple script would suffice. I wouldn't be surprised if there already is a script for this somewhere, as I think this is a rather common situation to occur.

[the remark this answers is not in the document]
Comment (andyboeh) [reply target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts:241]: The script I linked to in an earlier discussion does exactly that and works fine with `sysupgrade -n -F`
